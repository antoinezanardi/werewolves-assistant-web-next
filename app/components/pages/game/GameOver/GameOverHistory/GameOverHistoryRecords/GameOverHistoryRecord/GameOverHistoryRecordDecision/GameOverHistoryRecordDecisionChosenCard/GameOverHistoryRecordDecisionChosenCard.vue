<template>
  <div
    id="game-over-history-record-decision-chosen-card"
    class="flex flex-col gap-4 items-center justify-center"
  >
    <h4
      v-if="chosenCardText"
      id="game-over-history-record-decision-chosen-card-text"
      class="text-center"
    >
      {{ chosenCardText }}
    </h4>

    <RoleImage
      v-if="props.gameHistoryRecord.play.chosenCard"
      id="game-over-history-record-decision-chosen-card-role-image"
      v-tilt
      :role-name="props.gameHistoryRecord.play.chosenCard.roleName"
      :sizes="roleImageSizesInPx"
    />
  </div>
</template>

<script setup lang="ts">
import { GAME_HISTORY_RECORD_ROLE_IMAGE_SIZE_BELOW_MD, GAME_HISTORY_RECORD_ROLE_IMAGE_SIZE_OVER_MD } from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/game-over-history-record.constants";
import type { GameOverHistoryRecordDecisionChosenCardProps } from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/GameOverHistoryRecordDecision/GameOverHistoryRecordDecisionChosenCard/game-over-history-record-decision-chosen-card.types";
import RoleImage from "~/components/shared/role/RoleImage/RoleImage.vue";
import { useRoleName } from "~/composables/api/role/useRoleName.js";
import { useAppBreakpoints } from "~/composables/style/useAppBreakpoints";

const props = defineProps<GameOverHistoryRecordDecisionChosenCardProps>();
const { getDefiniteRoleNameLabel } = useRoleName();

const { isSmallerThanMdBreakpoint } = useAppBreakpoints();

const chosenCardText = computed<string>(() => {
  if (!props.gameHistoryRecord.play.chosenCard) {
    return "";
  }
  return getDefiniteRoleNameLabel(props.gameHistoryRecord.play.chosenCard.roleName, 1);
});

const roleImageSizesInPx = computed<string>(() => {
  if (isSmallerThanMdBreakpoint.value) {
    return `${GAME_HISTORY_RECORD_ROLE_IMAGE_SIZE_BELOW_MD}px`;
  }
  return `${GAME_HISTORY_RECORD_ROLE_IMAGE_SIZE_OVER_MD}px`;
});
</script>