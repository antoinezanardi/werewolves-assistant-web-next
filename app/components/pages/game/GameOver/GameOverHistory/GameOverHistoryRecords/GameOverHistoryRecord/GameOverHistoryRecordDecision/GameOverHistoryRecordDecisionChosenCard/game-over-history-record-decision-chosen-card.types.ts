import type { GameHistoryRecord } from "~/composables/api/game/types/game-history-record/game-history-record.class";

type GameOverHistoryRecordDecisionChosenCardProps = {
  gameHistoryRecord: GameHistoryRecord;
};

export type { GameOverHistoryRecordDecisionChosenCardProps };