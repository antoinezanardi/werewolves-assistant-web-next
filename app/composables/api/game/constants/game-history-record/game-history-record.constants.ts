const GAME_HISTORY_RECORD_PLAY_VOTING_RESULTS = [
  "sheriff-election",
  "tie",
  "death",
  "inconsequential",
  "skipped",
] as const;

const GAME_HISTORY_RECORD_PLAYER_ATTRIBUTE_ALTERATION_STATUSES = [
  "attached",
  "detached",
  "activated",
] as const;

export {
  GAME_HISTORY_RECORD_PLAY_VOTING_RESULTS,
  GAME_HISTORY_RECORD_PLAYER_ATTRIBUTE_ALTERATION_STATUSES,
};