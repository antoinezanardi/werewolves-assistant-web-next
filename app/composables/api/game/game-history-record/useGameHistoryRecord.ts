import type { GameHistoryRecord } from "~/composables/api/game/types/game-history-record/game-history-record.class";

type UseGameHistoryRecord = {
  didSourceSkipped: (gameHistoryRecord: GameHistoryRecord) => boolean;
};

function useGameHistoryRecord(): UseGameHistoryRecord {
  function didSourceSkipped(gameHistoryRecord: GameHistoryRecord): boolean {
    const {
      type,
      votes,
      targets,
      didJudgeRequestAnotherVote,
      chosenCard,
    } = gameHistoryRecord.play;

    return type === "vote" && !votes ||
      type === "target" && !targets ||
      type === "request-another-vote" && didJudgeRequestAnotherVote === undefined ||
      type === "choose-card" && !chosenCard;
  }
  return { didSourceSkipped };
}

export { useGameHistoryRecord };