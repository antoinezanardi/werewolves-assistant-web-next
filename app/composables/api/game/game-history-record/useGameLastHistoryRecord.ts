import type { MaybeRef, Ref } from "vue";
import type { GameHistoryRecordPlayerAttributeAlterationStatus } from "~/composables/api/game/types/game-history-record/game-history-record-player-attribute-alteration/game-history-record-player-attribute-alteration.types";
import type { GameHistoryRecord } from "~/composables/api/game/types/game-history-record/game-history-record.class";
import type { Game } from "~/composables/api/game/types/game.class";
import type { GameSource } from "~/composables/api/game/types/game.types";
import type { PlayerAttributeName } from "~/composables/api/game/types/players/player-attribute/player-attribute.types";
import type { Player } from "~/composables/api/game/types/players/player.class";

type UseGameLastHistoryRecord = {
  lastTargetedPlayers: Ref<Player[]>;
  doesHavePlayerAttributeAlteration: (attributeName: PlayerAttributeName, source: GameSource, status: GameHistoryRecordPlayerAttributeAlterationStatus) => boolean;
};

function useGameLastHistoryRecord(game: MaybeRef<Game>): UseGameLastHistoryRecord {
  const lastGameHistoryRecord = computed<GameHistoryRecord | null>(() => (isRef(game) ? game.value.lastGameHistoryRecord : game.lastGameHistoryRecord));

  const lastTargetedPlayers = computed<Player[]>(() => {
    if (!lastGameHistoryRecord.value?.play.targets) {
      return [];
    }
    return lastGameHistoryRecord.value.play.targets.map(target => target.player);
  });

  function doesHavePlayerAttributeAlteration(attributeName: PlayerAttributeName, source: GameSource, status: GameHistoryRecordPlayerAttributeAlterationStatus): boolean {
    if (!lastGameHistoryRecord.value?.playerAttributeAlterations) {
      return false;
    }
    return lastGameHistoryRecord.value.playerAttributeAlterations.some(playerAttributeAlteration => playerAttributeAlteration.name === attributeName &&
      playerAttributeAlteration.source === source &&
      playerAttributeAlteration.status === status);
  }
  return {
    lastTargetedPlayers,
    doesHavePlayerAttributeAlteration,
  };
}

export { useGameLastHistoryRecord };