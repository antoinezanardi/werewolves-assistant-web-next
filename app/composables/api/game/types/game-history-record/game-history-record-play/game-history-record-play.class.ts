import { Expose, plainToInstance, Type } from "class-transformer";

import { GameAdditionalCard } from "~/composables/api/game/types/game-additional-card/game-additional-card.class";
import { GameHistoryRecordPlaySource } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-source/game-history-record-play-source.class";
import { GameHistoryRecordPlayTarget } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-target/game-history-record-play-target.class";
import { GameHistoryRecordPlayVote } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-vote/game-history-record-play-vote.class";
import { GameHistoryRecordPlayVoting } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-voting/game-history-record-play-voting.class";
import type { GamePlayAction, GamePlayCause, GamePlayType } from "~/composables/api/game/types/game-play/game-play.types";
import type { RoleSide } from "~/composables/api/role/types/role.types";
import { DEFAULT_PLAIN_TO_INSTANCE_OPTIONS } from "~/utils/constants/class-transformer.constants";

class GameHistoryRecordPlay {
  @Expose()
  public type: GamePlayType;

  @Expose()
  public action: GamePlayAction;

  @Expose()
  @Type(() => GameHistoryRecordPlaySource)
  public source: GameHistoryRecordPlaySource;

  @Expose()
  public cause?: GamePlayCause;

  @Expose()
  @Type(() => GameHistoryRecordPlayTarget)
  public targets?: GameHistoryRecordPlayTarget[];

  @Expose()
  @Type(() => GameHistoryRecordPlayVote)
  public votes?: GameHistoryRecordPlayVote[];

  @Expose()
  @Type(() => GameHistoryRecordPlayVoting)
  public voting?: GameHistoryRecordPlayVoting;

  @Expose()
  public didJudgeRequestAnotherVote?: boolean;

  @Expose()
  @Type(() => GameAdditionalCard)
  public chosenCard?: GameAdditionalCard;

  @Expose()
  public chosenSide?: RoleSide;

  public static create(gameHistoryRecordPlay: GameHistoryRecordPlay): GameHistoryRecordPlay {
    return plainToInstance(GameHistoryRecordPlay, gameHistoryRecordPlay, DEFAULT_PLAIN_TO_INSTANCE_OPTIONS);
  }
}

export { GameHistoryRecordPlay };