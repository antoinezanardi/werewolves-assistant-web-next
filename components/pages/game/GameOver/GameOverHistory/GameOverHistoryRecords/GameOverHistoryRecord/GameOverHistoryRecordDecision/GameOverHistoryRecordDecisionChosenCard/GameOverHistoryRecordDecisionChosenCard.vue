<template>
  <div id="game-over-history-record-decision-chosen-card"/>
</template>