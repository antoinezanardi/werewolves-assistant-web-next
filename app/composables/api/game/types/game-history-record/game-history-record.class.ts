import { Expose, plainToInstance, Type } from "class-transformer";

import { GameHistoryRecordPlay } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play.class";
import { GameHistoryRecordPlayerAttributeAlteration } from "~/composables/api/game/types/game-history-record/game-history-record-player-attribute-alteration/game-history-record-player-attribute-alteration.class";
import { GamePhase } from "~/composables/api/game/types/game-phase/game-phase.class";
import { Player } from "~/composables/api/game/types/players/player.class";
import { DEFAULT_PLAIN_TO_INSTANCE_OPTIONS } from "~/utils/constants/class-transformer.constants";

class GameHistoryRecord {
  @Expose()
  public _id: string;

  @Expose()
  public gameId: string;

  @Expose()
  public turn: number;

  @Type(() => GamePhase)
  @Expose()
  public phase: GamePhase;

  @Expose()
  public tick: number;

  @Expose()
  @Type(() => GameHistoryRecordPlay)
  public play: GameHistoryRecordPlay;

  @Expose()
  @Type(() => Player)
  public revealedPlayers?: Player[];

  @Expose()
  @Type(() => Player)
  public deadPlayers?: Player[];

  @Expose()
  @Type(() => GameHistoryRecordPlayerAttributeAlteration)
  public playerAttributeAlterations?: GameHistoryRecordPlayerAttributeAlteration[];

  @Expose()
  public createdAt: Date;

  public static create(gameHistoryRecord: GameHistoryRecord): GameHistoryRecord {
    return plainToInstance(GameHistoryRecord, gameHistoryRecord, DEFAULT_PLAIN_TO_INSTANCE_OPTIONS);
  }
}

export { GameHistoryRecord };