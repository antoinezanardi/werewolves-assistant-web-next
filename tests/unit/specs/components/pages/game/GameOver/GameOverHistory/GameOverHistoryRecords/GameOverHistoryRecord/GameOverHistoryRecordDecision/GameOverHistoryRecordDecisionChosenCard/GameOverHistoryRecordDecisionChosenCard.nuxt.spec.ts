import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";
import type VueUseCore from "@vueuse/core";

import { createFakeGameAdditionalCard } from "@tests/unit/utils/factories/composables/api/game/game-additional-card/game-additional-card.factory";
import { createFakeGameHistoryRecordPlay } from "@tests/unit/utils/factories/composables/api/game/game-history-record/game-history-record-play/game-history-record-play.factory";
import { createFakeGameHistoryRecord } from "@tests/unit/utils/factories/composables/api/game/game-history-record/game-history-record.factory";
import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";
import type { GameOverHistoryRecordDecisionChosenCardProps } from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/GameOverHistoryRecordDecision/GameOverHistoryRecordDecisionChosenCard/game-over-history-record-decision-chosen-card.types";
import GameOverHistoryRecordDecisionChosenCard from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/GameOverHistoryRecordDecision/GameOverHistoryRecordDecisionChosenCard/GameOverHistoryRecordDecisionChosenCard.vue";
import type RoleImage from "~/components/shared/role/RoleImage/RoleImage.vue";

const hoistedMocks = vi.hoisted(() => ({
  useBreakpoints: {
    smaller: vi.fn(),
  },
}));

vi.mock("@vueuse/core", async importOriginal => ({
  ...await importOriginal<typeof VueUseCore>(),
  useBreakpoints: (): typeof hoistedMocks.useBreakpoints => hoistedMocks.useBreakpoints,
}));

describe("Game Over History Record Decision Chosen Card Component", () => {
  const defaultChosenCard = createFakeGameAdditionalCard({
    roleName: "seer",
  });
  const defaultGameHistoryRecord = createFakeGameHistoryRecord({
    play: createFakeGameHistoryRecordPlay({
      chosenCard: defaultChosenCard,
    }),
  });
  const defaultProps: GameOverHistoryRecordDecisionChosenCardProps = {
    gameHistoryRecord: defaultGameHistoryRecord,
  };
  let wrapper: ReturnType<typeof mount<typeof GameOverHistoryRecordDecisionChosenCard>>;

  async function mountGameOverHistoryRecordDecisionChosenCardComponent(options: ComponentMountingOptions<typeof GameOverHistoryRecordDecisionChosenCard> = {}):
  Promise<ReturnType<typeof mount<typeof GameOverHistoryRecordDecisionChosenCard>>> {
    return mountSuspendedComponent(GameOverHistoryRecordDecisionChosenCard, {
      shallow: false,
      global: {
        stubs: {
          RoleImage: false,
        },
      },
      props: defaultProps,
      ...options,
    });
  }

  beforeEach(async() => {
    hoistedMocks.useBreakpoints.smaller.mockReturnValue(ref(false));
    wrapper = await mountGameOverHistoryRecordDecisionChosenCardComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Chosen Card Text", () => {
    it("should translate chosen card text when rendered.", () => {
      const chosenCardText = wrapper.find<HTMLHeadingElement>("#game-over-history-record-decision-chosen-card-text");

      expect(chosenCardText.text()).toBe("shared.role.definiteName.seer, 1");
    });
  });

  describe("Chosen Card Role Image", () => {
    it("should pass chosen card role name to the role image component when rendered.", () => {
      const roleImage = wrapper.findComponent<typeof RoleImage>("#game-over-history-record-decision-chosen-card-role-image");

      expect(roleImage.props("roleName")).toBe(defaultChosenCard.roleName);
    });

    it("should pass 70px as the size to the role image component when rendered.", () => {
      const roleImage = wrapper.findComponent<typeof RoleImage>("#game-over-history-record-decision-chosen-card-role-image");

      expect(roleImage.props("sizes")).toBe("70px");
    });

    it("should pass 56px as the size to the role image component when the screen is smaller than md.", async() => {
      hoistedMocks.useBreakpoints.smaller.mockReturnValue(ref(true));
      wrapper = await mountGameOverHistoryRecordDecisionChosenCardComponent();
      const roleImage = wrapper.findComponent<typeof RoleImage>("#game-over-history-record-decision-chosen-card-role-image");

      expect(roleImage.props("sizes")).toBe("56px");
    });
  });
});