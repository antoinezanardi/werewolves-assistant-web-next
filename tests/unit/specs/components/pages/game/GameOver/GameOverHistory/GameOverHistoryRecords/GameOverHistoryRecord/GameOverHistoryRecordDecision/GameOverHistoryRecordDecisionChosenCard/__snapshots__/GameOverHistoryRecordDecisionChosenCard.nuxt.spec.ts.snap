// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Over History Record Decision Chosen Card Component > should match snapshot when rendered. 1`] = `
"<div id="game-over-history-record-decision-chosen-card" class="flex flex-col gap-4 items-center justify-center">
  <h4 id="game-over-history-record-decision-chosen-card-text" class="text-center">shared.role.definiteName.seer, 1</h4><img class="w-[70pxpx] h-[70pxpx] border-4 border-gray-800 role-image rounded-lg h-[70pxpx] w-[70pxpx] border-4 border-gray-800 role-image rounded-lg h-[70pxpx] w-[70pxpx]" width="70" height="70" alt="shared.role.name.seer" data-nuxt-img="" id="game-over-history-record-decision-chosen-card-role-image" src="/svg/misc/ripples.svg" style="will-change: transform; transition: 300ms cubic-bezier(.03,.98,.52,.99); transform: perspective(1000px) rotateX(NaNdeg) rotateY(NaNdeg) scale3d(1, 1, 1);">
</div>"
`;
